File: frontend/src/api/usage.ts
import client from "./client";
import type { UsageClient } from "../types";

export interface BrowserInfo {
  user_agent: string;
  browser_language: string;
  platform: string;
  timezone: string;
  screen_width: number;
  screen_height: number;
  viewport_width: number;
  viewport_height: number;
  current_path: string;
  referrer: string;
}

export interface ActiveUsersResponse {
  active_users: number;
  historical_users: number;
  window_seconds: number;
}

export interface AdminUsageSummaryResponse {
  active_users: number;
  historical_users: number;
  window_seconds: number;
}

export interface AdminUsageClientsResponse {
  clients: UsageClient[];
}

export function collectBrowserInfo(): BrowserInfo {
  return {
    user_agent: navigator.userAgent || "unknown",
    browser_language: navigator.language || "unknown",
    platform: navigator.platform || "unknown",
    timezone: Intl.DateTimeFormat().resolvedOptions().timeZone || "unknown",
    screen_width: window.screen?.width ?? 0,
    screen_height: window.screen?.height ?? 0,
    viewport_width: window.innerWidth ?? 0,
    viewport_height: window.innerHeight ?? 0,
    current_path: `${window.location.pathname}${window.location.search}` || "/",
    referrer: document.referrer || "",
  };
}

export async function sendUsageHeartbeat(
  clientId: string,
  browserInfo: BrowserInfo
): Promise<ActiveUsersResponse> {
  const { data } = await client.post<ActiveUsersResponse>("/usage/heartbeat", {
    client_id: clientId,
    browser_info: browserInfo,
  });
  return data;
}

export async function fetchAdminUsageSummary(token: string): Promise<AdminUsageSummaryResponse> {
  const { data } = await client.get<AdminUsageSummaryResponse>("/usage/admin/summary", {
    headers: { Authorization: `Bearer ${token}` },
  });
  return data;
}

export async function fetchAdminUsageClients(token: string): Promise<AdminUsageClientsResponse> {
  const { data } = await client.get<AdminUsageClientsResponse>("/usage/admin/clients", {
    headers: { Authorization: `Bearer ${token}` },
  });
  return data;
}

File: backend/app/schemas/usage.py
from pydantic import BaseModel, Field


class BrowserInfo(BaseModel):
    user_agent: str = Field(min_length=1, max_length=512)
    browser_language: str = Field(min_length=1, max_length=32)
    platform: str = Field(min_length=1, max_length=64)
    timezone: str = Field(min_length=1, max_length=64)
    screen_width: int = Field(ge=0)
    screen_height: int = Field(ge=0)
    viewport_width: int = Field(ge=0)
    viewport_height: int = Field(ge=0)
    current_path: str = Field(min_length=1, max_length=255)
    referrer: str = Field(default="", max_length=512)


class HeartbeatRequest(BaseModel):
    client_id: str = Field(min_length=8, max_length=64)
    browser_info: BrowserInfo


class ActiveUsersResponse(BaseModel):
    active_users: int
    historical_users: int
    window_seconds: int


class ActiveClientItem(BaseModel):
    client_id: str
    first_seen_at: str
    last_seen_at: str
    user_agent: str | None
    browser_language: str | None
    platform: str | None
    timezone: str | None
    screen_width: int | None
    screen_height: int | None
    viewport_width: int | None
    viewport_height: int | None
    current_path: str | None
    referrer: str | None
    ip_address: str | None


class AdminUsageSummaryResponse(BaseModel):
    active_users: int
    historical_users: int
    window_seconds: int


class AdminUsageClientsResponse(BaseModel):
    clients: list[ActiveClientItem]
